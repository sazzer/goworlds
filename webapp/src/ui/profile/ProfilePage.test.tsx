import React from 'react';
import configureStore from 'redux-mock-store';
import {render} from '@testing-library/react';
import {StaticRouter} from "react-router";
import {ProfilePage} from './index';
import {Provider} from "react-redux";

jest.mock('./ProfileHeader', () => {
    const ProfileHeader = jest.fn((props) => 'ProfileHeader: ' + JSON.stringify(props));
    return {
        ProfileHeader,
    };
});

jest.mock('./ProfileBreadcrumbs', () => {
    const ProfileBreadcrumbs = jest.fn((props) => 'ProfileBreadcrumbs: ' + JSON.stringify(props));
    return {
        ProfileBreadcrumbs,
    };
});

jest.mock('./ProfileForm', () => {
    const ProfileForm = jest.fn((props) => 'ProfileForm: ' + JSON.stringify(props));
    return {
        ProfileForm,
    };
});

jest.mock('./PasswordForm', () => {
    const PasswordForm = jest.fn((props) => 'PasswordForm: ' + JSON.stringify(props));
    return {
        PasswordForm,
    };
});

/** Set up the component to test */
function setup({state, location} : {state: any, location: string}) {
    const mockStoreCreator = configureStore();
    const store = mockStoreCreator(state);

    return {
        store,

        element: render(<StaticRouter location={location}><Provider store={store}><ProfilePage /></Provider></StaticRouter>)
    };
}

describe('Rendering the page', () => {
    describe('When the user is still loading', () => {
        const state = {
            currentUserId: {
                userId: 'abc123',
            },
            users: {
                users: [],
            },
        };

        it('renders the profile tab', () => {
            const {element, store} = setup({state, location: '/profile'});

            expect(element.baseElement).toMatchSnapshot();
            expect(store.getActions()).toHaveLength(1);
            expect(store.getActions()).toContainEqual(
                {
                    'type': 'Users/loadUser',
                    'payload': {
                        'userId': 'abc123',
                        'force': true,
                    },
                }
            )
        });

        it('renders the password tab', () => {
            const {element, store} = setup({state, location: '/profile/password'});

            expect(element.baseElement).toMatchSnapshot();
            expect(store.getActions()).toHaveLength(1);
            expect(store.getActions()).toContainEqual(
                {
                    'type': 'Users/loadUser',
                    'payload': {
                        'userId': 'abc123',
                        'force': true,
                    },
                }
            )
        });
    });

    describe('When the user is loaded', () => {
        const state = {
            currentUserId: {
                userId: 'abc123',
            },
            users: {
                users: [
                    {
                        id: 'abc123',
                        name: 'Graham',
                        email: 'graham@grahamcox.co.uk',
                        created: 'now',
                        updated: 'now',
                    }
                ],
            },
        };

        it('renders the profile tab', () => {
            const {element, store} = setup({state, location: '/profile'});

            expect(element.baseElement).toMatchSnapshot();
            expect(store.getActions()).toHaveLength(1);
            expect(store.getActions()).toContainEqual(
                {
                    'type': 'Users/loadUser',
                    'payload': {
                        'userId': 'abc123',
                        'force': true,
                    },
                }
            )
        });

        it('renders the password tab', () => {
            const {element, store} = setup({state, location: '/profile/password'});

            expect(element.baseElement).toMatchSnapshot();
            expect(store.getActions()).toHaveLength(1);
            expect(store.getActions()).toContainEqual(
                {
                    'type': 'Users/loadUser',
                    'payload': {
                        'userId': 'abc123',
                        'force': true,
                    },
                }
            )
        });
    });

});
