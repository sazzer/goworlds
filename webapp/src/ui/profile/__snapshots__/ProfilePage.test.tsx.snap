// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the page When the user is loaded renders the password tab 1`] = `
<body>
  <div>
    <div
      class="ui container"
    >
      ProfileBreadcrumbs: {"user":{"id":"abc123","name":"Graham","email":"graham@grahamcox.co.uk","created":"now","updated":"now"}}
      ProfileHeader: {"user":{"id":"abc123","name":"Graham","email":"graham@grahamcox.co.uk","created":"now","updated":"now"}}
      <div
        class="ui grid"
      >
        <div
          class="four wide column"
        >
          <div
            class="ui fluid pointing vertical menu"
          >
            <a
              class="item"
              href="/profile"
            >
              Profile
            </a>
            <a
              aria-current="page"
              class="item active"
              href="/profile/password"
            >
              Change Password
            </a>
          </div>
        </div>
        <div
          class="stretched twelve wide column"
        >
          PasswordForm: {"user":{"id":"abc123","name":"Graham","email":"graham@grahamcox.co.uk","created":"now","updated":"now"}}
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Rendering the page When the user is loaded renders the profile tab 1`] = `
<body>
  <div>
    <div
      class="ui container"
    >
      ProfileBreadcrumbs: {"user":{"id":"abc123","name":"Graham","email":"graham@grahamcox.co.uk","created":"now","updated":"now"}}
      ProfileHeader: {"user":{"id":"abc123","name":"Graham","email":"graham@grahamcox.co.uk","created":"now","updated":"now"}}
      <div
        class="ui grid"
      >
        <div
          class="four wide column"
        >
          <div
            class="ui fluid pointing vertical menu"
          >
            <a
              aria-current="page"
              class="item active"
              href="/profile"
            >
              Profile
            </a>
            <a
              class="item"
              href="/profile/password"
            >
              Change Password
            </a>
          </div>
        </div>
        <div
          class="stretched twelve wide column"
        >
          ProfileForm: {"userId":"abc123"}
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Rendering the page When the user is still loading renders the password tab 1`] = `
<body>
  <div>
    <div
      class="ui container"
    >
      ProfileBreadcrumbs: {"user":{"name":"Loading..."}}
      ProfileHeader: {"user":{"name":"Loading..."}}
      <div
        class="ui grid"
      >
        <div
          class="four wide column"
        >
          <div
            class="ui fluid pointing vertical menu"
          >
            <a
              class="item"
              href="/profile"
            >
              Profile
            </a>
            <a
              aria-current="page"
              class="item active"
              href="/profile/password"
            >
              Change Password
            </a>
          </div>
        </div>
        <div
          class="stretched twelve wide column"
        >
          <div
            class="ui active loader"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Rendering the page When the user is still loading renders the profile tab 1`] = `
<body>
  <div>
    <div
      class="ui container"
    >
      ProfileBreadcrumbs: {"user":{"name":"Loading..."}}
      ProfileHeader: {"user":{"name":"Loading..."}}
      <div
        class="ui grid"
      >
        <div
          class="four wide column"
        >
          <div
            class="ui fluid pointing vertical menu"
          >
            <a
              aria-current="page"
              class="item active"
              href="/profile"
            >
              Profile
            </a>
            <a
              class="item"
              href="/profile/password"
            >
              Change Password
            </a>
          </div>
        </div>
        <div
          class="stretched twelve wide column"
        >
          <div
            class="ui active loader"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
